src/pulsar_producer.erl

```diff
@@ -49,233 +90,579 @@ callback_mode() -> [state_functions].
+
+queue_item_sizer(?Q_ITEM(_CallId, _Ts, _Batch) = Item) ->
+    erlang:external_size(Item).
+
+queue_item_marshaller(?Q_ITEM(_, _, _) = I) ->
+  term_to_binary(I);
+queue_item_marshaller(Bin) when is_binary(Bin) ->
+  binary_to_term(Bin).
+
+now_ts() ->
+    erlang:system_time(millisecond).
+
+make_queue_item(From, Messages) ->
+    ?Q_ITEM(From, now_ts(), Messages).
+
+enqueue_send_requests(Requests, State = #{replayq := Q}) ->
+    QItems = lists:map(
+               fun(?SEND_REQ(From, Messages)) ->
+                 make_queue_item(From, Messages)
+               end,
+               Requests),
+    NewQ = replayq:append(Q, QItems),
+    State#{replayq := NewQ}.
+
+maybe_send_to_pulsar(State0) ->
+    #{ batch_size := BatchSize
+     , sequence_id := SequenceId
+     , requests := Requests0
+     , replayq := Q
+     , opts := ProducerOpts
+     } = State0,
+    case replayq:count(Q) =:= 0 of
+        true ->
+            State0;
+        false ->
+            MaxBatchBytes = maps:get(max_batch_bytes, ProducerOpts, ?DEFAULT_MAX_BATCH_BYTES),
+            {NewQ, QAckRef, Items} = replayq:pop(Q, #{ count_limit => BatchSize
+                                                     , bytes_limit => MaxBatchBytes
+                                                     }),
+            State1 = State0#{replayq := NewQ},
+            RetentionPeriod = maps:get(retention_period, ProducerOpts, infinity),
+            Now = now_ts(),
+            {Froms, Messages} =
+                lists:foldr(
+                  fun(?Q_ITEM(From, Timestamp, Msgs), {Froms, AccMsgs}) ->
+                    case is_batch_expired(Timestamp, RetentionPeriod, Now) of
+                      true -> {Froms, AccMsgs};
+                      false -> {[From | Froms], [{Timestamp, Msgs} | AccMsgs]}
+                    end
+                  end,
+                  {[], []},
+                  Items),
+            case Messages of
+                [] ->
+                    %% all expired, immediately ack replayq batch and continue
+                    ok = replayq:ack(Q, QAckRef),
+                    maybe_send_to_pulsar(State1);
+                [_ | _] ->
+                    send_batch_payload([Msg || {_Timestamp, Msgs} <- Messages, Msg <- Msgs],
+                                       SequenceId, State0),
+                    Requests = Requests0#{SequenceId => {QAckRef, Froms, Messages}},
+                    State2 = State1#{requests := Requests},
+                    State = next_sequence_id(State2),
+                    maybe_send_to_pulsar(State)
+            end
+    end.
 
-next_request_id(State = #state{request_id = ?MAX_QUE_ID}) ->
-    State#state{request_id = 1};
-next_request_id(State = #state{request_id = RequestId}) ->
-    State#state{request_id = RequestId+1}.
+collect_send_requests(Acc, Limit) ->
+    Count = 0,
+    do_collect_send_requests(Acc, Count, Limit).
+
+do_collect_send_requests(Acc, Count, Limit) when Count >= Limit ->
+    lists:reverse(Acc);
```

Comment:
the default value for `batch_size` config seems to be `0`
when Limit is zero, it will never make any progress ?

Reply:
ah, sorry, `Acc` is never empty.
we should maybe add an assertion here (since renaming the config will break compatibility)